### PR TITLE
RooChebychev: scalable implementation

### PR DESCRIPTION
**Pull request by manuelschiller**
use a more scalable implementation for RooChebychev which allows use of
Chebychev polynomial of arbitrarily high orders; also the extra effort
for every new order used is constant, so the effort (CPU) for evaluating
a RooChebychev grows linearly with the highest order Chebychev order
used, and comes in at only a couple of FLOPS per order (getting the
value of the coefficients is likely much more expensive)

FMA is used if supported by the underlying architecture/compiler,
otherwise the code falls back to normal floating point arithmetic.

RooChebychev is used in the stressRooFit test, which continues to run just fine.

### CODE REVIEW

**Comment by hageboeck** [comment]
Hey @manuelschiller,

that looks like a brilliant idea, thanks! We might get into trouble with the pragmas, though, because not all versions of ROOT are built with clang. We will see that when the bots get a chance to build it.

Would it be possible to write a test that compares the old and the new RooChebychev? A naive one would be to copy the old class into roofit/roofit/test/testRooChebychev.cxx, rename it, and just run the two against each other.
You might have tested locally, so including that would also be a good option.

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by manuelschiller** [comment]
> Hey @manuelschiller,
> 
> that looks like a brilliant idea, thanks! We might get into trouble with the pragmas, though, because not all versions of ROOT are built with clang. We will see that when the bots get a chance to build it.
> 
> Would it be possible to write a test that compares the old and the new RooChebychev? A naive one would be to copy the old class into roofit/roofit/test/testRooChebychev.cxx, rename it, and just run the two against each other.
> You might have tested locally, so including that would also be a good option.

As I said in my original commit message: This code is tested by stressRooFit (where a RooChebychev is used as a background model), so if this runs through fine (and it does), it means that the new code produces results that are indistinguishable from the reference file. Is that good enough?

Regarding the pragmas: Would it help if I tried hiding them behind some preprocessor magic? (According to https://en.cppreference.com/w/cpp/preprocessor/impl, unknown pragmas should be ignored by compilers.

**Comment by manuelschiller** [comment]
I also see that I need to be pure C++11. I'll amend later today.

**Comment by hageboeck** [comment]
> As I said in my original commit message: This code is tested by stressRooFit (where a RooChebychev is used as a background model), so if this runs through fine (and it does), it means that the new code produces results that are indistinguishable from the reference file. Is that good enough?

I see that now, yes.

File: roofit/roofit/src/RooChebychev.cxx [hunk 1]
```diff
@@ -34,17 +38,76 @@ Chebychev polynomial p.d.f. of the first kind
...
+ * @author Manuel Schiller <Manuel.Schiller@glasgow.ac.uk>
+ * @date 2019-03-24
+ */
+template <typename T, Kind KIND>
+class ChebychevIterator {
+private:
+   T m_last = 1;
+   T m_twox = 0;
+   T m_curr = 0;
+
```
Comment (hageboeck) [inline roofit/roofit/src/RooChebychev.cxx]: The ChebychevIterator is a really good idea. Could you change the member names fit into RooFit's naming convention? E.g. `_last`.

**Comment by manuelschiller** [comment]
> I took the liberty to get the member names in line with the RooFit conventions. I also added tags to typeset the formulas as latex.
> 
> Currently, there are problems with the build system, so we will have to wait a bit for the final test.

Thanks - and my apologies about the member names - I should have done that from the start, but it's hard to act against one's reflexes. Please let me know if there's anything else I should be doing.

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by hageboeck** [comment]
Suggest to squash on merge. The commit message could look like:
```
[RF] Extend RooChebychev to arbitrary orders.

Instead of hard-coding the polynomials to 7th order, a more scalable implementation
based on the recursive definition of the polynomials is used.
The computation time grows linearly with the order of the polynomials.

FMA is used if supported by the underlying architecture/compiler.
Otherwise, the code falls back to normal floating point arithmetic.
```

**Comment by manuelschiller** [comment]
Thanks all!

**Comment by hageboeck** [comment]
Actually, thank you!

**Pull request merged**